framework/app.mk

```diff
@@ -305,10 +306,40 @@ $(app_test_LIB): $(app_HEADER) $(app_plugin_deps) $(depend_libs) $(app_test_obje
 	@$(libmesh_LIBTOOL) --mode=install --quiet install -c $@ $(curr_dir)/lib
 endif
 
+# For static compiles, there are no references to symbols in the module, etc. object archives outside of
+# themselves.  This causes linkers to omit the symbols from those archives in the final binary
+# which prevents all of the moose objects to not be included/registered in the final binary.  To
+# force this we need to ad special linker flags to static builds.  Dynamic builds are immune since
+# the linker doesn't know what symbols will be needed by any binary that will dynamically link to
+# it in the future - so all symbols remain intact in the dylib/so in that case and are present and
+# registered as expected.  See https://stackoverflow.com/questions/5202142/static-variable-initialization-over-a-library
+# and https://stackoverflow.com/questions/9459980/c-global-variable-not-initialized-when-linked-through-static-libraries-but-ok#11336506
+# for more explanations/details.
+applibs := $(app_LIBS)
+applibs += $(app_test_LIB)
+ifneq ($(libmesh_shared),yes)
+  ifneq (,$(findstring clang,$(CXX)))
+    # replace .a with .dylib for testing in dynamic configuration:
+    applibs := $(foreach lib,$(patsubst %.la,%.a,$(applibs)),-Wl,-force_load,$(lib))
+  else
+  ifneq (,$(findstring g++,$(CXX)))
+    applibs := $(foreach lib,$(patsubst %.la,%.a,$(applibs)),-Wl,--whole-archive,$(lib),--no-whole-archive)
+  else
+  ifneq (,$(findstring mpicxx,$(CXX)))
```

Comment:
It seems to me that you should find clang/gcc (and what about icc) outside of this logic and then do the if tests so that you don't need this duplication for the mpicxx case.

Reply:
Yeah - I'll take a swing at it.